enhancements/update/change-management-and-maintenance-schedules.md

```diff
@@ -0,0 +1,1334 @@
+
+  Examples:
+  - Adds a finalizer to namespaces. Namespace cannot be deleted without our controller running.
+  - Restricts the label format for objects to X.
+  - Defaults field Y on object kind Z.
+
+Fill in the operational impact of these API Extensions in the "Operational Aspects
+of API Extensions" section.
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+In the HCP topology, the `HostedCluster` and `NodePool` resources are enhanced to support the `spec.changeManagement`
+stanza. 
+
+#### Standalone Clusters
+
+In the Standalone topology, the ClusterVersion and MachineConfigPool resources are enhanced to support the 
+`spec.changeManagement` stanza.
+
+#### Single-node Deployments or MicroShift
+
+These toplogies do not have worker nodes. Only the ClusterVersion change management policy will be relevant.
+There is no logical distinction between user workloads and control-plane workloads in this case. A control-plane
+update will drain user workloads and will cause workload disruption.
+
+Though disruption is inevitable, the maintenance schedule feature provides value in these toplogies by
+giving the cluster lifecycle administrator high level controls on exactly when it occurs.
+
+#### OCM Managed Profiles
+OpenShift Cluster Manager (OCM) should expose a user interface allowing users to manage their change management policy.
+Standard Fleet clusters will expose the option to configure the control-plane and worker-node `ChangeManagementPolicy`
+objects with the `MaintenanceSchedule` strategy - including permit and exclude times.
+
+- Service Delivery will reserve the right to disable this policy for emergency corrective actions.
+- Service Delivery should constrain permit & exclude configurations based on their internal policies. 
+  For example, customers may be forced to enable permissive windows which amount to at least 6 hours a month.
+
+### Implementation Details/Notes/Constraints
+
+#### ChangeManagement Stanza
+The `spec.changeManagement` stanza will be introduced into ClusterVersion and MachineConfigPool (for standalone profiles)
+and HostedCluster and NodePool (for HCP profiles).
+
+#### MaintenanceSchedule Strategy Configuration
+
+When a `ChangeManagementPolicy` is defined to use the `MaintenanceSchedule` strategy,
+a `maintenanceSchedule` stanza should also be provided to configure the strategy
+(if it is not, the policy is functionally identical to `Restrictive`).
+
+```yaml
+kind: ChanageManagementPolicy
+spec:
+  strategy: MaintenanceSchedule
+ 
+  maintenanceSchedule:
+    # Specifies a reoccurring permissive window. 
+    permit:  
+      # RRULEs (https://www.rfc-editor.org/rfc/rfc5545#section-3.3.10) are commonly used 
+      # for calendar management metadata. Only a subset of the RFC is supported.
+      # See "RRULE Constraints" section for details.
+      # If unset, all dates are permitted and only exclude constrains permissive windows.
+      recurrence: <rrule|null>

```

Comment:
Still think we want to make a structured API that we convert into RRULE btw, it will be more native and I think we can make it intuitive and have an easier time validating the user intent and restricting what we do and don't allow, rather than exposing the entirety of RRULE

Reply:
Thinking about this API a bit more, I'm not sure exactly where we want to draw the lines of what we do and don't support, but I could imagine something along the lines of a couple of layers of discriminated unions.

In particular, the structure below would mean that all fields are required (depending on discriminators), so any missing or incomplete information to work out an accurate rule would be easily relayed to the end user, and, we have a specific intent from them (ie RRULE=Yearly implying 1st day of the year, would have to be explicit here, ie a more declarative API)

We would also be able to specify individual limits on each of the intervals and appropriate doc text to explain what the interval means at each level.
```
recurrence:
  frequency: Yearly | Monthly | Weekly | Daily
  daily:
    interval: <int> # How many days should pass before the next occurrence 0 < x < 31
  weekly:
    days: [Monday, Tuesday, Wednesday, ...]
    interval: <int> # How many weeks should pass before the next occurrence 0 < x < 27
  monthly:
    by: Day | Date
    date:
      dates: 
      - <int> # List of dates 0 < x < 32
      interval: <int> # How many months should pass before the next occurrence 0 < x < 12
    day:
      days:
      - week: First | Second | Third | Fourth | Fifth | Last
        day: Monday | Tuesday | Wednesday | ... | Sunday
      interval: <int> # How many months should pass before the next occurrence 0 < x < 12
  yearly:
    by: Day | Date
    date:
      dates: 
      - <int> # List of dates 0 < x < 32
      month: January | February | ... | December
    day:
      days:
      - week: First | Second | Third | Fourth | Fifth | Last
        day: Monday | Tuesday | Wednesday | ... | Sunday
      month: January | February | ... | December
```

Some real world examples

```
# Every second day
recurrence:
  frequency: Daily
  daily:
    interval: 2

# Every Saturday and Sunday
recurrence:
  frequency: Weekly
  weekly:
    days:
    - Saturday
    - Sunday
    interval: 1

# The last Monday of each month
recurrence:
  frequency: Monthly
  monthly:
    by: Day
    day:
      days:
      - day: Monday
        week: Last
      interval: 1
```

It is more verbose, but we can document and validate the user intention more easily, without caveats of "If you don't specify XYZ, this means ABC"

Reply:
Thanks @JoelSpeed - The latest push includes the schema for folks to review.

Reply:
{code}
As RRULE: `FREQ=MONTHLY;BYDAY=MO;BYSETPOS=-1`
Through enhancement schema:
```yaml
recurrence:
  frequency: Monthly
  monthly:
    by: Day
    day:
      days:
      - day: Monday
        week: Last
      interval: 1
{code}

Is there a way to cut the redundant use of the word `day`? When I look at this, I can't help seeing  `Day day days day`

Reply:
Sadly, as is the nature of very declarative APIs and discriminated unions in particular.

Perhaps being a bit more explicit about the names, `on` instead of `days`, `dayOfWeek` and `weekOfMonth` to be more explicit about what these mean
```
recurrence:
  frequency: Monthly
  monthly:
    by: Day
    day:
      on:
      - dayOfWeek: Monday
        weekOfMonth: Last
      interval: 1
```

Reply:
yeah, that sounds way better to me, it's easier to understand